### PR TITLE
Warp Up  Homework

### PR DESCRIPTION
**Pull request by victornani07**

### CODE REVIEW

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 1]
```diff
@@ -21,7 +30,13 @@ public boolean noneMatch(final int[] input) {
      */
     @Override
     public boolean someMatch(final int[] input, IntPredicate predicate) {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        for(int num : input) {
+            if (predicate.test(num)) {
+                return true;
+            }
+        }
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: delete this line

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 2]
```diff
@@ -34,7 +49,15 @@ public boolean someMatch(final int[] input, IntPredicate predicate) {
...
                             IntPredicate predicate) {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        for(String s : input) {
+            int num = function.applyAsInt(s);
+
+            if(!predicate.test(num)) {
+                return false;
+            }
+        }
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: this line also delete

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 3]
```diff
@@ -34,7 +49,15 @@ public boolean someMatch(final int[] input, IntPredicate predicate) {
     public boolean allMatch(final String[] input,
                             ToIntFunction<String> function,
                             IntPredicate predicate) {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        for(String s : input) {
+            int num = function.applyAsInt(s);
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: Here also there is not need of a empty line, here is not so much code to be not enough readable

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 4]
```diff
@@ -46,7 +69,12 @@ public boolean allMatch(final String[] input,
      */
     @Override
     public int[] copyValues(int[] input, int startInclusive, int endExclusive) throws IllegalArgumentException {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        if(startInclusive < 0 || endExclusive >= input.length || startInclusive >= endExclusive) {
+            throw new IllegalArgumentException();
+        }
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: remove the empty line

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 5]
```diff
@@ -46,7 +69,12 @@ public boolean allMatch(final String[] input,
      */
     @Override
     public int[] copyValues(int[] input, int startInclusive, int endExclusive) throws IllegalArgumentException {
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java:66]: Recommend here to use the try-catch blocks (throwing the Exception in the catch block)

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 6]
```diff
@@ -76,7 +111,21 @@ public int findSecondMax(final int[] input) {
...
+                .count();
+
+        for(int k = input.length - 1; k >= 0; --k) {
+            if (input[k] < 0) {
+                newValues[i++] = input[k];
+            } else {
+                newValues[j++] = input[k];
+            }
+        }
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: you can delete this line

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 7]
```diff
@@ -99,7 +152,24 @@ public int findSecondMax(final int[] input) {
      */
     @Override
     public int[] insertValues(final int[] input, int startInclusive, int[] values) throws IllegalArgumentException {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        if(startInclusive < 0 || startInclusive >= input.length) {
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: here also you must use try-catch block

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 8]
```diff
@@ -99,7 +152,24 @@ public int findSecondMax(final int[] input) {
...
+        for(int i = 0; i < input.length; ++i) {
+            if(i == startInclusive) {
+                while (j < values.length) {
+                    newValues[i + j] = values[j++];
+                }
+            }
+
+            newValues[i + j] = input[i];
+        }
+
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java:168]: remove empty line

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 9]
```diff
@@ -111,7 +181,44 @@ public int findSecondMax(final int[] input) {
      */
     @Override
     public int[] mergeSortedArrays(int[] input, int[] input2) throws IllegalArgumentException {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        final int N = input.length;
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: give more descriptive names for variables/constants. Constants must be declared as final and static, and their names must be only with upper case.

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 10]
```diff
@@ -111,7 +181,44 @@ public int findSecondMax(final int[] input) {
...
+        final int M = input2.length;
+        int[] newValues = new int[N + M];
+        int i = 0, j = 0, k = 0;
+
+        while(i < N && j < M) {
+            if(input[i] < input2[j]) {
+                if(i > 0 && input[i] < input[i - 1]) {
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: remove all verifications like these from everywhere, add a single try-catch block, there in the try block will be all code, and in catch will be "throw new IllegalArgumentException()"

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 11]
```diff
@@ -139,7 +284,20 @@ public void validateForMatrixMultiplication(int[][] leftMatrix, int[][] rightMat
      */
     @Override
     public int[][] matrixMultiplication(final int[][] leftMatrix, final int[][] rightMatrix) throws NullPointerException, IllegalArgumentException {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        validateForMatrixMultiplication(leftMatrix, rightMatrix);
+
+        int M = leftMatrix.length, N = rightMatrix[0].length;
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: here also rename constants and make them final static

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 12]
```diff
@@ -148,6 +306,12 @@ public void validateForMatrixMultiplication(int[][] leftMatrix, int[][] rightMat
      */
     @Override
     public int[] distinct(final int[] input) {
-        throw new UnsupportedOperationException("Method not implemented yet");
+        Set<Integer> s = Arrays.stream(input)
```
Comment (dspeianu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: There is a method distinct() that you can use over a stream to get distinct values.

File: src/test/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoopsTest.java [hunk 13]
```diff
@@ -197,6 +197,7 @@ void distinct() {
             new int[3]
         };
         int[][] rightMatrix1 = {
+                new int[5],
```
Comment (dspeianu) [inline src/test/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoopsTest.java]: reformat code here

**Review by dspeianu — CHANGES_REQUESTED** [review]

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 14]
```diff
@@ -99,7 +146,25 @@ public int findSecondMax(final int[] input) {
...
+                if (i == startInclusive) {
+                    while (j < values.length) {
+                        newValues[i + j] = values[j++];
+                    }
+                }
+                newValues[i + j] = input[i];
+            }
+            return newValues;
+        } catch (Exception e) {
+            throw new IllegalArgumentException();
```
Comment (victornani07) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java:166]: This should have been: "e.printStackTrace()". My bad.

**Review by dspeianu — APPROVED** [review]

**Pull request merged**